File: content/snippets/scala3/interview_lcci/02/04_partition_list.scala
object Lcci0204PartitionList:
  final case class ListNode(var value: Int, var next: ListNode | Null = null)

  def buildList(nums: Vector[Int]): ListNode | Null =
    val dummy = ListNode(0, null)
    var tail: ListNode = dummy
    nums.foreach { n =>
      tail.next = ListNode(n, null)
      tail = tail.next.asInstanceOf[ListNode]
    }
    dummy.next

  def partition(head: ListNode | Null, x: Int): ListNode | Null =
    val lessDummy = ListNode(0, null)
    val greaterDummy = ListNode(0, null)
    var lessTail: ListNode = lessDummy
    var greaterTail: ListNode = greaterDummy
    var cur = head

    while (cur != null) {
      val node = cur.asInstanceOf[ListNode]
      if (node.value < x) {
        lessTail.next = node
        lessTail = node
      } else {
        greaterTail.next = node
        greaterTail = node
      }
      cur = node.next
    }

    greaterTail.next = null
    lessTail.next = greaterDummy.next
    lessDummy.next

  def printList(head: ListNode | Null): Unit =
    val out = scala.collection.mutable.ArrayBuffer.empty[Int]
    var cur = head
    while (cur != null) {
      val node = cur.asInstanceOf[ListNode]
      out += node.value
      cur = node.next
    }
    println(out.mkString(" "))

  @main def main(): Unit =
    val head = buildList(Vector(3, 5, 8, 5, 10, 2, 1))
    printList(partition(head, 5))

File: content/snippets/typescript/interview_lcci/02/04_partition_list.ts
class ListNode {
  val: number;
  next: ListNode | null;

  constructor(val: number, next: ListNode | null = null) {
    this.val = val;
    this.next = next;
  }
}

function buildList(nums: number[]): ListNode | null {
  const dummy = new ListNode(0);
  let tail = dummy;
  for (const n of nums) {
    tail.next = new ListNode(n);
    tail = tail.next;
  }
  return dummy.next;
}

function partition(head: ListNode | null, x: number): ListNode | null {
  const lessDummy = new ListNode(0);
  const greaterDummy = new ListNode(0);
  let lessTail = lessDummy;
  let greaterTail = greaterDummy;

  while (head !== null) {
    if (head.val < x) {
      lessTail.next = head;
      lessTail = lessTail.next;
    } else {
      greaterTail.next = head;
      greaterTail = greaterTail.next;
    }
    head = head.next;
  }

  greaterTail.next = null;
  lessTail.next = greaterDummy.next;
  return lessDummy.next;
}

function printList(head: ListNode | null): void {
  const out: number[] = [];
  while (head !== null) {
    out.push(head.val);
    head = head.next;
  }
  console.log(out.join(" "));
}

function main(): void {
  const head = buildList([3, 5, 8, 5, 10, 2, 1]);
  printList(partition(head, 5));
}

main();
